### PR TITLE
Unify internal property creation

### PR DESCRIPTION
**Pull request by zherczeg**

### CODE REVIEW

File: jerry-core/ecma/base/ecma-globals.h [hunk 1]
```diff
@@ -614,11 +554,55 @@ typedef struct
...
+
+/**
+ * Checks whether a property is internal property
+ */
+#define ECMA_PROPERTY_IS_INTERNAL(property) ((property) >= ECMA_PROPERTY_INTERNAL)
+
+/**
+ * Checks whether a property is raw data or accessor property
+ */
+#define ECMA_PROPERTY_IS_RAW(property) \
```
Comment (lygstate) [inline jerry-core/ecma/base/ecma-globals.h:571]: I like something like (property & 0b00011000)
Without the need of compare operator
Any definition of property bits? 
property is 8 bits long, what's the meaning of each bits?

Comment (zherczeg) [reply jerry-core/ecma/base/ecma-globals.h:571]: Good bit magic is essential for high performance. This condition checks that the highest two bits are not set *without* using a temporary register. The alternative ´(x & 0xc0) != 0xc0´ requires two operations and a temporary register.
Property bits are described in ecma_property_flags_t.

**Review by rerobika — APPROVED** [review]
LGTM

**Comment by LaszloLango** [comment]
Property creation and usage are hot paths in the engine. Does this patch have any effect on the performance?

**Review by LaszloLango — CHANGES_REQUESTED** [review]
Please update `./docs/04.INTERNALS.md` as well.

**Comment by zherczeg** [comment]
Benchmark | Heap (bytes) | Stack (bytes) | Perf (sec) |
----: | ----: | ----: | ----: | 
3d-cube.js | 10992 -> 10992 : 0.000% | 14232 -> 14232 : 0.000% | 1.132 -> 1.122 : +0.850% | 
3d-morph.js | 289232 -> 289232 : 0.000% | 1704 -> 1704 : 0.000% | 1.318 -> 1.324 : -0.449% | 
3d-raytrace.js | 117656 -> 117656 : 0.000% | 2596 -> 2596 : 0.000% | 1.326 -> 1.320 : +0.525% | 
access-binary-trees.js | 27120 -> 27120 : 0.000% | 3512 -> 3512 : 0.000% | 0.778 -> 0.790 : -1.574% | 
access-fannkuch.js | 1792 -> 1792 : 0.000% | 1776 -> 1776 : 0.000% | 2.546 -> 2.555 : -0.354% | 
access-nbody.js | 4600 -> 4600 : 0.000% | 1864 -> 1864 : 0.000% | 1.519 -> 1.521 : -0.119% | 
bitops-3bit-bits-in-byte.js | 1024 -> 1024 : 0.000% | 1720 -> 1720 : 0.000% | 0.703 -> 0.701 : +0.291% | 
bitops-bits-in-byte.js | 984 -> 984 : 0.000% | 1720 -> 1720 : 0.000% | 1.015 -> 1.013 : +0.247% | 
bitops-bitwise-and.js | 752 -> 752 : 0.000% | 1472 -> 1472 : 0.000% | 1.251 -> 1.238 : **+1.060%** | 
bitops-nsieve-bits.js | 61184 -> 61184 : 0.000% | 1720 -> 1720 : 0.000% | 1.477 -> 1.476 : +0.094% | 
controlflow-recursive.js | 1288 -> 1288 : 0.000% | 63624 -> 63624 : 0.000% | 0.520 -> 0.523 : -0.458% | 
crypto-aes.js | 28040 -> 28040 : 0.000% | 2312 -> 2312 : 0.000% | 0.884 -> 0.887 : -0.367% | 
crypto-md5.js | 70968 -> 70968 : 0.000% | 1788 -> 1788 : 0.000% | 0.742 -> 0.742 : +0.104% | 
crypto-sha1.js | 46328 -> 46328 : 0.000% | 1848 -> 1848 : 0.000% | 0.764 -> 0.758 : +0.707% | 
crypto.js | 51704 -> 51704 : 0.000% | 3228 -> 3228 : 0.000% | 8.684 -> 8.657 : +0.307% | 
date-format-tofte.js | 8952 -> 8952 : 0.000% | 2120 -> 2120 : 0.000% | 1.012 -> 1.004 : +0.760% | 
date-format-xparb.js | 15144 -> 15144 : 0.000% | 2704 -> 2704 : 0.000% | 0.820 -> 0.813 : +0.854% | 
deltablue.js | 455544 -> 455544 : 0.000% | 3636 -> 3636 : 0.000% | 5.487 -> 5.465 : +0.403% | 
math-cordic.js | 2048 -> 2048 : 0.000% | 1720 -> 1720 : 0.000% | 1.711 -> 1.709 : +0.098% | 
math-partial-sums.js | 1512 -> 1512 : 0.000% | 1696 -> 1696 : 0.000% | 1.154 -> 1.156 : -0.144% | 
math-spectral-norm.js | 3264 -> 3264 : 0.000% | 1720 -> 1720 : 0.000% | 0.739 -> 0.739 : +0.073% | 
raytrace.js | 22424 -> 22424 : 0.000% | 5104 -> 5104 : 0.000% | 2.774 -> 2.791 : -0.597% | 
richards.js | 8568 -> 8568 : 0.000% | 2196 -> 2196 : 0.000% | 0.314 -> 0.314 : +0.093% | 
string-base64.js | 90104 -> 90104 : 0.000% | 1768 -> 1768 : 0.000% | 1.434 -> 1.420 : **+1.038%** | 
string-fasta.js | 4400 -> 4400 : 0.000% | 1704 -> 1704 : 0.000% | 3.420 -> 3.425 : -0.133% | 
Geometric mean: | 0.000% | 0.000% | +0.134% | 

Binary (bytes) | master(99ec5b9bab) | patch(136fa184ed) | Diff |
----: | ----: | ----: | ----: |
size | 226976 | 226976 | 0 bytes |
.rodata | 20003 | 20003 | 0 bytes |
.dynstr | 448 | 448 | 0 bytes |
.rel.plt | 456 | 456 | 0 bytes |
.interp | 25 | 25 | 0 bytes |
.dynsym | 1008 | 1008 | 0 bytes |
.gnu.hash | 464 | 464 | 0 bytes |
.text | 198352 | 198312 | -40 bytes |
.comment | 85 | 85 | 0 bytes |
.shstrtab | 226 | 226 | 0 bytes |
.data | 8 | 8 | 0 bytes |
.ARM.exidx | 8 | 8 | 0 bytes |
.rel.dyn | 16 | 16 | 0 bytes |
.init | 12 | 12 | 0 bytes |
.got | 244 | 244 | 0 bytes |
.plt | 716 | 716 | 0 bytes |
.note.ABI-tag | 32 | 32 | 0 bytes |
.gnu.version_r | 64 | 64 | 0 bytes |
.bss | 1575088 | 1575088 | 0 bytes |
.fini | 8 | 8 | 0 bytes |
.hash | 408 | 408 | 0 bytes |
.gnu.version | 126 | 126 | 0 bytes |
.fini_array | 4 | 4 | 0 bytes |
.init_array | 4 | 4 | 0 bytes |
.dynamic | 248 | 248 | 0 bytes |
.eh_frame | 4 | 4 | 0 bytes |
.ARM.attributes | 53 | 53 | 0 bytes |

No major effect.

**Review by LaszloLango — APPROVED** [review]
LGTM

File: jerry-core/ecma/operations/ecma-objects.c [hunk 2]
```diff
@@ -2413,18 +2411,14 @@ ecma_op_object_own_property_keys (ecma_object_t *obj_p) /**< object */
...
-            && prop_pair_p->names_cp[i] >= LIT_NON_INTERNAL_MAGIC_STRING__COUNT
-            && prop_pair_p->names_cp[i] < LIT_MAGIC_STRING__COUNT)
-        {
-          /* Internal properties are never enumerated. */
-          continue;
-        }
+        /* Internal properties are special properties so they must not be present. */
+        JERRY_ASSERT (ECMA_PROPERTY_GET_NAME_TYPE (*property_p) != ECMA_DIRECT_STRING_MAGIC
+                      || prop_pair_p->names_cp[i] < LIT_NON_INTERNAL_MAGIC_STRING__COUNT
+                      || prop_pair_p->names_cp[i] >= LIT_MAGIC_STRING__COUNT);
```
Comment (ossy-szeged) [inline jerry-core/ecma/operations/ecma-objects.c:2421]: This assert hits on debug test262-esnext on 116 tests. Please a take a look what happened here.

**Pull request merged**